compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -618,17 +622,34 @@ private boolean isAssignable(BType source, BType target, Set<TypePair> unresolve
                 isArrayTypesAssignable(source, target, unresolvedTypes);
     }
 
-    private boolean isAssignableRecordType(BRecordType recordType, BMapType targetMapType) {
+    private boolean isAssignableRecordType(BRecordType recordType, BType type) {
+        BType targetType;
+        switch (type.tag) {
+            case TypeTags.MAP:
+                targetType = ((BMapType) type).constraint;
+                break;
+            case TypeTags.JSON:
+                targetType = type;
+                break;
+            default:
+                throw new IllegalArgumentException("Incompatible target type: " + type.toString());
+        }
+
         if (recordType.sealed) {
-            return recordFieldsAssignableToMap(recordType, targetMapType);
+            return recordFieldsAssignableToType(recordType, targetType);
         } else {
-            return isAssignable(recordType.restFieldType, targetMapType.constraint)
-                    && recordFieldsAssignableToMap(recordType, targetMapType);
+            return recordFieldsAssignableToType(recordType, targetType) &&
+                    isAssignable(recordType.restFieldType, targetType);
```

Comment:
Shall we move this logic to `recordFieldsAssignableToType`. Rest fields are also record fields?

Reply:
+1.. That'd actually simplify things a bit